bitbots_vision/src/bitbots_vision/dynamic_color_space.py

```diff
@@ -62,71 +68,49 @@ def vision_config_callback(self, msg):
         :return: None
         """
         # Load dict from string in yaml-format in msg.data
-        vision_config = yaml.load(msg.data)
+        vision_config = yaml.load(msg.data, Loader=yaml.FullLoader)
 
-        self.debug_printer = debug.DebugPrinter(
-            debug_classes=debug.DebugPrinter.generate_debug_class_list_from_string(
-                vision_config['vision_debug_printer_classes']))
         self.runtime_evaluator = evaluator.RuntimeEvaluator(None)
 
         # Print status of dynamic color space after toggeling 'dynamic_color_space_active' parameter
-        if 'dynamic_color_space_active' not in self.vision_config or \
-            vision_config['dynamic_color_space_active'] != self.vision_config['dynamic_color_space_active']:
+        if ros_utils.ROS_Utils.config_param_change(self.vision_config, vision_config, 'dynamic_color_space_active'):
             if vision_config['dynamic_color_space_active']:
                 rospy.loginfo('Dynamic color space turned ON.')
             else:
                 rospy.logwarn('Dynamic color space turned OFF.')
 
         # Set publisher of ColorSpace-messages
-        if 'ROS_dynamic_color_space_msg_topic' not in self.vision_config or \
-                self.vision_config['ROS_dynamic_color_space_msg_topic'] != vision_config['ROS_dynamic_color_space_msg_topic']:
-            if hasattr(self, 'pub_color_space'):
-                self.pub_color_space.unregister()
-            self.pub_color_space = rospy.Publisher(
-                vision_config['ROS_dynamic_color_space_msg_topic'],
-                ColorSpace,
-                queue_size=1)
+        self.pub_color_space = ros_utils.ROS_Utils.create_or_update_publisher(self.vision_config, vision_config, self.pub_color_space, 'ROS_dynamic_color_space_msg_topic', ColorSpace)
+
 
         # Set Color- and FieldBoundaryDetector
         self.color_detector = color.DynamicPixelListColorDetector(
-            self.debug_printer,
             self.package_path,
             vision_config)
 
-        self.field_boundary_detector = field_boundary.FieldBoundaryDetector(
+        # Get field boundary detector class by name from config
+        field_boundary_detector_class = field_boundary.FieldBoundaryDetector.get_by_name(
+            vision_config['dynamic_color_space_field_boundary_detector_search_method'])
+
+        # Set the field boundary detector
+        self.field_boundary_detector = field_boundary_detector_class(
             self.color_detector,
             vision_config,
-            self.debug_printer,
-            used_by_dyn_color_detector=True)
-
-        # Reset queue
-        if hasattr(self, 'color_value_queue'):
-            self.color_value_queue.clear()
```

Comment:
Removing these three lines might cause unexpected behavior, even tho the queue will be reassigned afterwards.
#EvaluateThis

Reply:
The parameter `used_by_dyn_color_detector` had only been used for the selection of the right detector. In the old implementation the field boundary detector looked in the config which detector type should be used. The dynamic colorspace had an own detector config. To determine which config parameter should be loaded, the constructor had this optional parameter to load the dynamic colorspace config.
In the new implementation with the class hierachy this isn't necessary anymore because the dynamic colorspace selects the right class depending on his config.
Of course the cleaning of the queue has to be evaluated. I remember some this line existed because of some strange reasons.